### PR TITLE
.Net: MS AI Ollama with Function calling

### LINKED ISSUE #7442: .Net: Ollama LocalModels FunctionCalling Support
Update Ollama Connector to allow usage of the pattern for function calling using the `Ollama’s raw mode.`

Reference:
- [Mistral FunctionCalling](https://ollama.com/library/mistral)

### PR DESCRIPTION
**Pull request by RogerBarreto**
# Problem & Solution

This PR updates the Ollama Connector to use it's latest `OllamaSharp` updated version `4.0` where already supports the new `Microsoft.Extensions.AI` that can be completely leveraged by SK Abstractions including Function Calling support.

- Resolves #7442

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.Ollama/Core/ServiceBase.cs [hunk 1]
```diff
@@ -59,4 +61,12 @@ internal ServiceBase(string model,
...
         this.AttributesInternal.Add(AIServiceExtensions.ModelIdKey, model);
     }
+
+    /// <summary>
+    /// Disposes the client.
+    /// </summary>
+    ~ServiceBase()
+    {
+        this._client.Dispose();
+    }
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Ollama/Core/ServiceBase.cs:71]: This isn't needed or desirable.

File: dotnet/src/Connectors/Connectors.Ollama/Extensions/OllamaKernelBuilderExtensions.cs [hunk 2]
```diff
@@ -143,19 +143,17 @@ public static IKernelBuilder AddOllamaChatCompletion(
...
     public static IKernelBuilder AddOllamaChatCompletion(
         this IKernelBuilder builder,
-        string modelId,
         OllamaApiClient ollamaClient,
         string? serviceId = null)
     {
         Verify.NotNull(builder);
 
-        builder.Services.AddOllamaChatCompletion(modelId, ollamaClient, serviceId);
+        builder.Services.AddOllamaChatCompletion(ollamaClient, serviceId);
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Ollama/Extensions/OllamaKernelBuilderExtensions.cs:156]: If you want to bake in a modelId, you can, e.g.
```C#
.UseChatOptions(options =>
{
    options ??= new();
    options.ModelId ??= modelId;
    return options;
})
```

File: dotnet/src/Connectors/Connectors.Ollama/Extensions/OllamaServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -110,10 +111,20 @@ public static IServiceCollection AddOllamaChatCompletion(
...
+            var ollamaClient = new OllamaApiClient(endpoint, modelId);
+            var logger = serviceProvider.GetService<ILoggerFactory>()?.CreateLogger(ollamaClient.GetType());
+
+            var chatClientBuilder = new ChatClientBuilder()
+                .UseFunctionInvocation();
+
+            if (logger is not null)
+            {
+                chatClientBuilder.UseLogging(logger);
+            }
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.Ollama/Extensions/OllamaServiceCollectionExtensions.cs:124]: Might want to use UseOpenTelemetry instead (or possibly in addition, though it'll be somewhat duplicative).

Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.Ollama/Extensions/OllamaServiceCollectionExtensions.cs:124]: Will keep logging and if there's asks, may consider adding a flag to use telemetry in the extension.

Comment (stephentoub) [reply dotnet/src/Connectors/Connectors.Ollama/Extensions/OllamaServiceCollectionExtensions.cs:124]: Doesn't that differ from the existing connectors?

File: dotnet/src/Connectors/Connectors.Ollama/Extensions/OllamaServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -110,10 +111,21 @@ public static IServiceCollection AddOllamaChatCompletion(
...
-                modelId: modelId,
-                endpoint: endpoint,
-                loggerFactory: serviceProvider.GetService<ILoggerFactory>()));
+        {
+            var ollamaClient = new OllamaApiClient(endpoint, modelId);
+            var logger = serviceProvider.GetService<ILoggerFactory>()?.CreateLogger(ollamaClient.GetType());
+
+            var chatClientBuilder = new ChatClientBuilder()
+                .UseFunctionInvocation(config =>
+                    config.MaximumIterationsPerRequest = 128);
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.Ollama/Extensions/OllamaServiceCollectionExtensions.cs:120]: Should this parameter be configurable? If not, then taking into account that it's used in multiple places, it's worth to define `128` as constant.

File: dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatCompletionServiceChatClient.cs [hunk 5]
```diff
@@ -40,7 +40,7 @@ public ChatCompletionServiceChatClient(IChatCompletionService chatCompletionServ
         Verify.NotNull(chatMessages);
 
         var response = await this._chatCompletionService.GetChatMessageContentAsync(
-            new ChatHistory(chatMessages.Select(ChatCompletionServiceExtensions.ToChatMessageContent)),
+            new ChatHistory(chatMessages.Select(m => ChatCompletionServiceExtensions.ToChatMessageContent(m))),
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatCompletionServiceChatClient.cs:43]: Is this change required?

File: dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatCompletionServiceChatClient.cs [hunk 6]
```diff
@@ -58,7 +58,7 @@ public async IAsyncEnumerable<StreamingChatCompletionUpdate> CompleteStreamingAs
         Verify.NotNull(chatMessages);
 
         await foreach (var update in this._chatCompletionService.GetStreamingChatMessageContentsAsync(
-            new ChatHistory(chatMessages.Select(ChatCompletionServiceExtensions.ToChatMessageContent)),
+            new ChatHistory(chatMessages.Select(m => ChatCompletionServiceExtensions.ToChatMessageContent(m))),
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatCompletionServiceChatClient.cs:61]: Is this change required?

File: dotnet/src/SemanticKernel.Abstractions/AI/Embeddings/EmbeddingGenerationServiceExtensions.cs [hunk 7]
```diff
@@ -159,4 +174,11 @@ public async Task<IList<ReadOnlyMemory<TEmbedding>>> GenerateEmbeddingsAsync(ILi
             return embeddings.Select(e => e.Vector).ToList();
         }
     }
+
+    private sealed class EmbeddingGeneratorTextEmbeddingGenerationService : EmbeddingGeneratorEmbeddingGenerationService<string, float>, ITextEmbeddingGenerationService
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel.Abstractions/AI/Embeddings/EmbeddingGenerationServiceExtensions.cs:178]: Is this class required?

**Review by dmytrostruk — APPROVED** [review]

File: dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatCompletionServiceChatClient.cs (hunk 5 above)
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatCompletionServiceChatClient.cs:43]: It is because the method accepts two parameters instead of one.

File: dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatCompletionServiceChatClient.cs (hunk 6 above)
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatCompletionServiceChatClient.cs:61]: Answered above

File: dotnet/src/SemanticKernel.Abstractions/AI/Embeddings/EmbeddingGenerationServiceExtensions.cs (hunk 7 above)
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Abstractions/AI/Embeddings/EmbeddingGenerationServiceExtensions.cs:178]: Yes, otherwise I won't be able  to cast `IEmbeddingGenerationService<T,Y>` to `ITextEmbeddingGenerationService`, I need a class that formally implements `ITextEmbeddingGenerationService`.

Comment (stephentoub) [reply dotnet/src/SemanticKernel.Abstractions/AI/Embeddings/EmbeddingGenerationServiceExtensions.cs:178]: You could change the implementation of AsEmbeddingGenerationService to return a type that also implements the interface; then this can just cast to it.

Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Abstractions/AI/Embeddings/EmbeddingGenerationServiceExtensions.cs:178]: I would have to put the `private sealed class` as public for it.

When I attempt to return `ITextGEnerationService` from this signature also feels a bit odd, as the return name is different from the method, that's why my first suggestion was to have just the `AsTextEmbeddingGenerationService` while this interface still exists.

Comment (stephentoub) [reply dotnet/src/SemanticKernel.Abstractions/AI/Embeddings/EmbeddingGenerationServiceExtensions.cs:178]: > I would have to put the private sealed class as public for it.

No, I mean you'd change this:
https://github.com/microsoft/semantic-kernel/blob/c539e8b88b4a30b7bc008bcdff2fd49336cd5096/dotnet/src/SemanticKernel.Abstractions/AI/Embeddings/EmbeddingGenerationServiceExtensions.cs#L121-L122
to be this:
```C#
    private sealed class EmbeddingGeneratorTextEmbeddingGenerationService(
        IEmbeddingGenerator<string, Embedding<float>> generator, IServiceProvider? serviceProvider) :
        EmbeddingGeneratorEmbeddingGenerationService<string, float>(generator, serviceProvider), ITextEmbeddingGenerationService
    { 
    }

    private class EmbeddingGeneratorEmbeddingGenerationService<TValue, TEmbedding> : IEmbeddingGenerationService<TValue, TEmbedding>
        where TEmbedding : unmanaged
```
and you'd change this:
https://github.com/microsoft/semantic-kernel/blob/60ff436ecd0975fbe1cd4745cd2d207078e27e2a/dotnet/src/SemanticKernel.Abstractions/AI/Embeddings/EmbeddingGenerationServiceExtensions.cs#L73-L75
to this:
```C#
        return 
            generator is IEmbeddingGenerationService<TValue, TEmbedding> service ? service :
            typeof(TValue) == typeof(string) && typeof(TEmbedding) == typeof(float) ? new EmbeddingGeneratorTextEmbeddingGenerationService(generator, serviceProvider) :
            new EmbeddingGeneratorEmbeddingGenerationService<TValue, TEmbedding>(generator, serviceProvider);
```
or something like that. That way, any time AsEmbeddingGenerationService is used with string/float, the resulting object also implements ITextEmbeddingGenerationService.

File: dotnet/src/Connectors/Connectors.Ollama/Extensions/OllamaServiceCollectionExtensions.cs [hunk 8]
```diff
@@ -135,10 +147,24 @@ public static IServiceCollection AddOllamaChatCompletion(
...
+                client: HttpClientProvider.GetHttpClient(httpClient, serviceProvider),
+                modelId);
+
+            var logger = serviceProvider.GetService<ILoggerFactory>()?.CreateLogger(ollamaClient.GetType());
+
+            var chatClientBuilder = new ChatClientBuilder()
+                .UseFunctionInvocation(config =>
+                    config.MaximumIterationsPerRequest = 128);
+
+            if (logger is not null)
```
Comment (ladeak) [inline dotnet/src/Connectors/Connectors.Ollama/Extensions/OllamaServiceCollectionExtensions.cs:161]: nitpicking: creating the logger (line 155), could be moved to "just" before it is being used (before the `if (logger is not null)`

File: dotnet/src/InternalUtilities/test/MultipleHttpMessageHandlerStub.cs [hunk 9]
```diff
@@ -52,4 +52,14 @@ protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage
 
         return await Task.FromResult(this.ResponsesToReturn[this._callIteration - 1]);
     }
+
+    internal string? GetRequestContentAsString(int index, Encoding? encoding = null)
+    {
+        if (encoding is null)
```
Comment (ladeak) [inline dotnet/src/InternalUtilities/test/MultipleHttpMessageHandlerStub.cs:58]: Could this use `encoding ??= Encoding.UTF8;`

File: dotnet/src/SemanticKernel.Abstractions/AI/Embeddings/EmbeddingGenerationServiceExtensions.cs (hunk 7 above)
Comment (RogerBarret0) [reply dotnet/src/SemanticKernel.Abstractions/AI/Embeddings/EmbeddingGenerationServiceExtensions.cs:178]: Opted for the `AsTextEmbeddingGenerationService`, as this is more simple and this method can be individually Obsoleted and removed easily in the future.

File: dotnet/src/Connectors/Connectors.Ollama/Extensions/OllamaServiceCollectionExtensions.cs (hunk 4 above)
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.Ollama/Extensions/OllamaServiceCollectionExtensions.cs]: Moved to a constant, similar to OpenAI.

**Review by SergeyMenshykh — APPROVED** [review]
